tensorflow/compiler/mlir/xla/transforms/lhlo_copy_removal.cc

```diff
@@ -0,0 +1,78 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// This file implements a pass to remove redundant LHLO copy operations.
+
+#include "absl/memory/memory.h"
+#include "mlir/Dialect/StandardOps/Ops.h"  // TF:llvm-project
+#include "mlir/IR/Operation.h"             // TF:llvm-project
+#include "mlir/Pass/Pass.h"                // TF:llvm-project
+#include "tensorflow/compiler/mlir/xla/ir/lhlo_ops.h"
+#include "tensorflow/compiler/mlir/xla/transforms/passes.h"
+
+namespace mlir {
+namespace xla_hlo {
+namespace {
+
+/// Removes Lhlo.CopyOp that copies from an allocated buffer to the block
+/// argument. All uses of the buffer are replaced with the block argument.
+struct LhloCopyRemoval : mlir::FunctionPass<LhloCopyRemoval> {
+  void runOnFunction() override {
+    llvm::SmallVector<mlir::Operation*, 2> eraseList;
+    getFunction().walk([&](mlir::xla_lhlo::CopyOp copyOp) {
+      mlir::Value fromOperand = copyOp.operand();
+      mlir::Value toOperand = copyOp.output();
+
+      // If the fromOperand value is a block argument or the toOperand
+      // value is not a block argument, then ignore this copy operation.
+      if (!fromOperand.getDefiningOp() || toOperand.getDefiningOp()) {
+        return;
+      }
+
+      // Remove the associated alloc operation.
+      auto allocOp = fromOperand.getDefiningOp();
+      eraseList.push_back(allocOp);
+
+      // Iterate over all uses of the fromOperand to find the associated
+      // deallocOp (if any).
+      for (auto op : fromOperand.getUsers()) {
+        if (isa<mlir::DeallocOp>(op)) {
+          eraseList.push_back(op);
+          break;
+        }
+      }
+
+      // Replace all uses of the fromOperand with the toOperand. This rewires
+      // all references pointing to the original alloc operation to the new
+      // target operation in order to safely remove the copy op.
+      fromOperand.replaceAllUsesWith(toOperand);
+      copyOp.erase();
+    });
```

Comment:
Seems like you're looking at:
```
^bb(%buffer):
  ...
  %src = op()
  ...
  produce(%src)
  ....
  copy(%src, %buffer)
  ...
  dealloc(%src)
  ...
```

And you want to have:

```
^bb(%buffer):
  ...
  produce(%buffer)
  ...
```

I think there is a missing validity criteria here because you're assuming that buffer isn't used after the source of the copy is produced, for example if the input looked like:

```
^bb(%buffer):
  ...
  %src = op()
  ...
  produce(%src)
  ...
  use(%buffer)
  ....
  copy(%src, %buffer)
  ...
  dealloc(%src)
  ...
```

A TODO to call this out later fix this detect and prevent replacement seems appropriate to me.

Another aspect to call out is the possible aliasing between the block arguments.

Reply:
Thanks for the feedback; we have been aware of these limitations. Our initial assumption was that this "pass" was tightly coupled with the buffer assignment that happens during HLO-to-LHLO-legalization. We totally agree that these cases could happen and we should try to create a generic version in the future that does not make IR-based assumptions on the underlying structure of the program. Hence, we added additional checks to avoid illegal removals in which we accidentally remove CopyOps where we should not remove them. In one of the next PRs, we are going to address the aliasing problems.